code/modules/mob/living/carbon/carbon.dm

```diff
@@ -911,18 +911,22 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, list(/obj/machinery/atmospherics/unary/ven
 	return FALSE
 
 /mob/living/carbon/resist_fire()
+	if(HAS_TRAIT(src, TRAIT_FIRE_ROLLING))
+		return
 	fire_stacks -= 5
-	Weaken(6 SECONDS, TRUE) //We dont check for CANWEAKEN, I don't care how immune to weakening you are, if you're rolling on the ground, you're busy.
+	ADD_TRAIT(src, TRAIT_FIRE_ROLLING, TRAIT_GENERIC)
+	KnockDown(6 SECONDS, TRUE) //We dont check for CANWEAKEN, I don't care how immune to weakening you are, if you're rolling on the ground, you're busy.
 	spin(32, 2)
 	visible_message("<span class='danger'>[src] rolls on the floor, trying to put [p_themselves()] out!</span>",
 		"<span class='notice'>You stop, drop, and roll!</span>")
-	sleep(3 SECONDS)
+	addtimer(CALLBACK(src, PROC_REF(extinguish_roll), 3 SECONDS))
+
+/mob/living/carbon/proc/extinguish_roll()
+	REMOVE_TRAIT(src, TRAIT_FIRE_ROLLING, TRAIT_GENERIC)
 	if(fire_stacks <= 0)
-		visible_message("<span class='danger'>[src] has successfully extinguished [p_themselves()]!</span>",
-			"<span class='notice'>You extinguish yourself.</span>")
+		visible_message("<span class='danger'>[src] has successfully extinguished [p_themselves()]!</span>","<span class='notice'>You extinguish yourself.</span>")
 		ExtinguishMob()
```

Comment:
Is this whole delayed part even actually necessary? `living/proc/handle_fire()` already calls `ExtinguishMob()` if your fire stacks go under 0 (and should always fire within the 3 seconds of delay); there's also the `adjust_fire_stacks()` proc which also auto-extinguishes at 0 fire stacks

Reply:
That can happen up to a cycle after this happens, its nice for the fire to go out at the same time that the extinguishing message appears.

Reply:
Cycles are generally 2 seconds, so it's guaranteed to happen before the timer's 3s delay
As for the message happening at the same time, I'm actually surprised to find out you're supposed to extinguish only 3s after pressing Resist, usually it happens instantaneously